# Objective

Build a model to automatically predict tags for a given a StackExchange question by using the text of the question.
![alt text](https://cdn.sstatic.net/Sites/stackoverflow/company/img/logos/se/se-logo.svg?v=d29f0785ebb7)

__Dataset Specs__: Over 85,000 questions

__License__

All Stack Exchange user contributions are licensed under [CC-BY-SA 3.0](https://creativecommons.org/licenses/by-sa/3.0/) with [attribution required](http://blog.stackoverflow.com/2009/06/attribution-required/).

<br>

***

In [1]:
# optional step (if you are working on colab)
# mount Google drive
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Steps to Follow
1. Loading Data
2. Text Cleaning
3. Merge Tags with Questions
4. Dataset Preparation
5. Feature Engineering using TF-IDF
6. Model Building
    1. Naive Bayes
    2. Logistic Regression
    3. Model Building Summary

7. Final Question Tagging Pipeline



In [2]:
# Importing Required Libraries

# for string matching
import re 

# for handling data
import pandas as pd

# for numerical computing
import numpy as np

# for handling html data
from bs4 import BeautifulSoup

# for NLP related tasks
import spacy
nlp=spacy.load('en_core_web_sm',disable=["tagger", "parser","ner"])

pd.set_option('display.max_colwidth', 200)

# Loading Data

In [3]:
# load questions
questions_df = pd.read_csv('/content/drive/My Drive/dataset/Questions.csv',encoding='latin-1')
print('Shape=>',questions_df.shape)
questions_df.head()

Shape=> (85085, 6)


Id  ...                                                                                                                                                                                                     Body
0   6  ...  <p>Last year, I read a blog post from <a href="http://anyall.org/">Brendan O'Connor</a> entitled <a href="http://anyall.org/blog/2008/12/statistics-vs-machine-learning-fight/">"Statistics vs. Mach...
1  21  ...  <p>What are some of the ways to forecast demographic census with some validation and calibration techniques?</p>\n\n<p>Some of the concerns:</p>\n\n<ul>\n<li>Census blocks vary in sizes as rural\n...
2  22  ...                                                                               <p>How would you describe in plain English the characteristics that distinguish Bayesian from Frequentist reasoning?</p>\n
3  31  ...  <p>After taking a statistics course and then trying to help fellow students, I noticed one subject that inspires much head-desk banging is interpreting the results of statistical hypothesis tests....
4  36  ...  <p>There is an old saying: "Correlation does not mean causation". When I teach, I tend to use the following standard examples to illustrate this point:</p>\n\n<ol>\n<li>number of storks and birth ...

[5 rows x 6 columns]

1. Id: Question ID
2. OwnerUserId: User ID
3. CreationDate: Date of posting question
4. Score: Count of Upvotes received by the question
5. Title: Title of the question
6. Body: Text body of the question

In [4]:
# load tags
tags_df = pd.read_csv('/content/drive/My Drive/dataset/Tags.csv')
print('Shape=>',tags_df.shape)
tags_df.head()

Shape=> (244228, 2)


Id            Tag
0   1       bayesian
1   1          prior
2   1    elicitation
3   2  distributions
4   2      normality

# Text Cleaning

Let's define a function to clean the text data.

In [5]:
def cleaner(text):

  # take off html tags
  text = BeautifulSoup(text).get_text()
  
  # fetch alphabetic characters
  text = re.sub("[^a-zA-Z]", " ", text)

  # convert text to lower case
  text = text.lower()
  
  # removing extra spaces
  text=re.sub("[\s]+"," ",text)

  # creating doc object
  doc=nlp(text)

  # remove stopwords and lemmatize the text
  tokens=[token.lemma_ for token in doc if(token.is_stop==False)]

  return " ".join(tokens)

In [6]:
# Pre-processing Questions
questions_df['cleaned_text'] = questions_df['Body'].apply(cleaner)

In [7]:
questions_df['Body'][1]

"<p>What are some of the ways to forecast demographic census with some validation and calibration techniques?</p>\n\n<p>Some of the concerns:</p>\n\n<ul>\n<li>Census blocks vary in sizes as rural\nareas are a lot larger than condensed\nurban areas. Is there a need to account for the area size difference?</li>\n<li>if let's say I have census data\ndating back to 4 - 5 census periods,\nhow far can i forecast it into the\nfuture?</li>\n<li>if some of the census zone change\nlightly in boundaries, how can i\naccount for that change?</li>\n<li>What are the methods to validate\ncensus forecasts? for example, if i\nhave data for existing 5 census\nperiods, should I model the first 3\nand test it on the latter two? or is\nthere another way?</li>\n<li>what's the state of practice in\nforecasting census data, and what are\nsome of the state of the art methods?</li>\n</ul>\n"

In [8]:
questions_df['cleaned_text'][1]

'way forecast demographic census validation calibration technique concern census block vary size rural area lot large condense urban area need account area size difference let s census datum date census period far forecast future census zone change lightly boundary account change method validate census forecast example datum exist census period model test way s state practice forecast census datum state art method'

# Merge Tags with Questions

Let's now explore the tags data.

In [9]:
tags_df.head()

Id            Tag
0   1       bayesian
1   1          prior
2   1    elicitation
3   2  distributions
4   2      normality

In [10]:
# count of unique tags
len(tags_df['Tag'].unique())

1315

In [11]:
tags_df['Tag'].value_counts()

r                   13236
regression          10959
machine-learning     6089
time-series          5559
probability          4217
                    ...  
mboost                  1
fda                     1
leave-one-out           1
dimensions              1
2d                      1
Name: Tag, Length: 1315, dtype: int64

In [12]:
# remove "-" from the tags
tags_df['Tag']= tags_df['Tag'].apply(lambda x:re.sub("-"," ",x))

In [13]:
# group tags Id wise
tags_df = tags_df.groupby('Id').apply(lambda x:x['Tag'].values).reset_index(name='tags')
tags_df.head()

Id                                       tags
0   1             [bayesian, prior, elicitation]
1   2                 [distributions, normality]
2   3                    [software, open source]
3   4  [distributions, statistical significance]
4   6                         [machine learning]

In [14]:
# merge tags and questions
df = pd.merge(questions_df,tags_df,how='inner',on='Id')

In [15]:
df = df[['Id','Body','cleaned_text','tags']]
print('Shape=>',df.shape)
df.head()

Shape=> (85085, 4)


Id  ...                                                              tags
0   6  ...                                                [machine learning]
1  21  ...                                 [forecasting, population, census]
2  22  ...                                           [bayesian, frequentist]
3  31  ...  [hypothesis testing, t test, p value, interpretation, intuition]
4  36  ...                                           [correlation, teaching]

[5 rows x 4 columns]

There are over 85,000 unique questions and over 1300 tags.

# Dataset Preparation

In [16]:
# check frequency of occurence of each tag
freq= {}
for i in df['tags']:
  for j in i:
    if j in freq.keys():
      freq[j] = freq[j] + 1
    else:
      freq[j] = 1

Let's find out the most frequent tags.

In [17]:
# sort the dictionary in descending order
freq = dict(sorted(freq.items(), key=lambda x:x[1],reverse=True))

In [18]:
freq.items()

dict_items([('r', 13236), ('regression', 10959), ('machine learning', 6089), ('time series', 5559), ('probability', 4217), ('hypothesis testing', 3869), ('self study', 3732), ('distributions', 3501), ('logistic', 3316), ('classification', 2881), ('correlation', 2871), ('statistical significance', 2666), ('bayesian', 2656), ('anova', 2505), ('normal distribution', 2181), ('multiple regression', 2054), ('mixed model', 1998), ('clustering', 1952), ('neural networks', 1897), ('mathematical statistics', 1888), ('confidence interval', 1776), ('categorical data', 1703), ('generalized linear model', 1614), ('variance', 1576), ('data visualization', 1549), ('estimation', 1533), ('forecasting', 1422), ('t test', 1418), ('pca', 1395), ('sampling', 1363), ('cross validation', 1344), ('repeated measures', 1335), ('spss', 1296), ('svm', 1283), ('chi squared', 1261), ('maximum likelihood', 1209), ('predictive models', 1189), ('multivariate analysis', 1116), ('survival', 1081), ('references', 1076), (

In [19]:
# Top 10 most frequent tags
common_tags = list(freq.keys())[:10]
common_tags

['r',
 'regression',
 'machine learning',
 'time series',
 'probability',
 'hypothesis testing',
 'self study',
 'distributions',
 'logistic',
 'classification']

We will use only those questions/queries that have the above 10 tags associated with it.

In [20]:
x=[]
y=[]

for i in range(len(df['tags'])):
  
  temp=[]
  for j in df['tags'][i]:
    if j in common_tags:
      temp.append(j)

  if(len(temp)>1):
    x.append(df['cleaned_text'][i])
    y.append(temp)

In [21]:
# number of questions left
len(x)

11106

In [22]:
x[:5]

['recently start work tuberculosis clinic meet periodically discuss numb tb case currently treat numb test administer etc have like start model count guess unusual unfortunately have little train time series exposure model continuous datum stock price large number count influenza deal case month mean median var distribute like image lose mist time image eat grue have find article address model like have greatly appreciate hear suggestion approach r package use implement approach edit mbq s answer force think carefully be ask get hang monthly count lose actual focus question have like know fairly visible decline onward reflect downward trend overall numb case look like numb case monthly reflect stable process maybe seasonality overall stable present look like process change overall numb case decline monthly count wobble randomness seasonality test s real change process identify decline use trend seasonality estimate numb case upcoming month whew thank bear',
 'look well behave distribut

In [23]:
y[:5]

[['r', 'time series'],
 ['regression', 'distributions'],
 ['distributions', 'probability', 'hypothesis testing'],
 ['hypothesis testing', 'self study'],
 ['r', 'regression', 'time series']]

In [24]:
from sklearn.preprocessing import MultiLabelBinarizer
mlb = MultiLabelBinarizer()

# Getting Labels 
y = mlb.fit_transform(y)
y.shape

(11106, 10)

In [25]:
y[0,:]

array([0, 0, 0, 0, 0, 0, 1, 0, 0, 1])

In [26]:
mlb.classes_

array(['classification', 'distributions', 'hypothesis testing',
       'logistic', 'machine learning', 'probability', 'r', 'regression',
       'self study', 'time series'], dtype=object)

We can now split the dataset into training set and validation set. 

In [27]:
from sklearn.model_selection import train_test_split
x_tr,x_val,y_tr,y_val=train_test_split(x, y, test_size=0.2, random_state=0,shuffle=True)

In [28]:
print('x_tr:',len(x_tr),'y_tr:',len(y_tr))
print('x_val:',len(x_val),'y_val:',len(y_val))

x_tr: 8884 y_tr: 8884
x_val: 2222 y_val: 2222


# Feature Engineering using TF-IDF

In [29]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [30]:
# initialize TFIDF
word_vectorizer = TfidfVectorizer(max_features=5000)

In [31]:
# Fitting Vectorizer on Train set
word_vectorizer.fit(x_tr)

TfidfVectorizer(analyzer='word', binary=False, decode_error='strict',
                dtype=<class 'numpy.float64'>, encoding='utf-8',
                input='content', lowercase=True, max_df=1.0, max_features=5000,
                min_df=1, ngram_range=(1, 1), norm='l2', preprocessor=None,
                smooth_idf=True, stop_words=None, strip_accents=None,
                sublinear_tf=False, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, use_idf=True, vocabulary=None)

In [32]:
# create TF-IDF vectors for Train Set
train_word_features = word_vectorizer.transform(x_tr)

In [33]:
train_word_features

<8884x5000 sparse matrix of type '<class 'numpy.float64'>'
	with 394812 stored elements in Compressed Sparse Row format>

In [34]:
# create TF-IDF vectors for Test Set
test_word_features = word_vectorizer.transform(x_val)
test_word_features

<2222x5000 sparse matrix of type '<class 'numpy.float64'>'
	with 96172 stored elements in Compressed Sparse Row format>

# Model Building

## Naive Bayes

In [35]:
# Importing for modeling
from sklearn.multiclass import OneVsRestClassifier
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import f1_score

In [36]:
# Defining Model
nb_model=OneVsRestClassifier(MultinomialNB())

In [37]:
# Training Model
nb_model.fit(train_word_features,y_tr)

OneVsRestClassifier(estimator=MultinomialNB(alpha=1.0, class_prior=None,
                                            fit_prior=True),
                    n_jobs=None)

In [38]:
# Make predictions for train set
train_pred_nb=nb_model.predict_proba(train_word_features)

In [39]:
train_pred_nb[:5]

array([[0.01263068, 0.02913076, 0.27116896, 0.01677616, 0.01916243,
        0.01913419, 0.83814495, 0.10788506, 0.05706641, 0.93865151],
       [0.01317187, 0.01175525, 0.03202753, 0.11629246, 0.02707439,
        0.01234429, 0.49022885, 0.77020863, 0.01645327, 0.540419  ],
       [0.04911206, 0.01735512, 0.0269517 , 0.25034582, 0.08523484,
        0.02020249, 0.34495066, 0.58351473, 0.0290121 , 0.02140273],
       [0.01715074, 0.01528334, 0.05202715, 0.02676137, 0.05508736,
        0.01800138, 0.47646249, 0.29866026, 0.06519084, 0.61215162],
       [0.00443262, 0.00101109, 0.00297146, 0.00520821, 0.03163826,
        0.00108445, 0.95240362, 0.12258498, 0.00207054, 0.99491953]])

The predictions are in terms of probabilities for each of the 10 tags. Hence we need to have a threshold value to convert these probabilities to 0 or 1.

Let's specify a set of candidate threshold values. We will select the threshold value that performs the best for the train set.

In [40]:
# Function for converting probabilities into classes or tags based on a threshold value
def classify(pred_prob,threshold):
  y_pred_seq = []

  for i in pred_prob:
    temp=[]
    for j in i:
      if j>=threshold:
        temp.append(1)
      else:
        temp.append(0)
    y_pred_seq.append(temp)

  return y_pred_seq

In [41]:
# Function for finding optimum value of threshold
def optimum_threshold(actual,pred_prob):
  #define candidate threshold values
  thresholds  = np.arange(0,0.5,0.01)

  score=[]
  for value in thresholds:
    # Getting classes for each threshold
    pred_classes= classify(pred_prob,value) 
    # Getting F1-score for every threshold
    score.append(f1_score(actual,pred_classes,average="weighted"))

  return thresholds[score.index(max(score))]    

In [42]:
# Finding Optimum value
print("Optimal threshold=>",optimum_threshold(y_tr,train_pred_nb))

Optimal threshold=> 0.26


In [43]:
# Getting classes using optimum threshold
train_pred_nb_class=classify(train_pred_nb,0.26)

In [44]:
train_pred_nb_class[:5]

[[0, 0, 1, 0, 0, 0, 1, 0, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 1, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 1, 0, 0],
 [0, 0, 0, 0, 0, 0, 1, 1, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 0, 0, 1]]

In [45]:
mlb.inverse_transform(np.array(train_pred_nb_class[:5]))

[('hypothesis testing', 'r', 'time series'),
 ('r', 'regression', 'time series'),
 ('r', 'regression'),
 ('r', 'regression', 'time series'),
 ('r', 'time series')]

In [46]:
# Evaluating on Training Set
print("F1-score on Train Set:",f1_score(y_tr,train_pred_nb_class,average="weighted"))

F1-score on Train Set: 0.7273434612826666


In [47]:
# Make Predictions on Validation Set
val_pred_nb=nb_model.predict_proba(test_word_features)

# Getting Classes
val_pred_nb_class=classify(val_pred_nb,0.26)

# Evaluating on Validation Set
print("F1-score on Validation Set:",f1_score(y_val,val_pred_nb_class,average="weighted"))

F1-score on Validation Set: 0.6622003021672971


## Logistic Regression

In [48]:
from sklearn.linear_model import LogisticRegression

In [49]:
# Defining Model
lr_model=OneVsRestClassifier(LogisticRegression())

# Training Model
lr_model.fit(train_word_features,y_tr)

OneVsRestClassifier(estimator=LogisticRegression(C=1.0, class_weight=None,
                                                 dual=False, fit_intercept=True,
                                                 intercept_scaling=1,
                                                 l1_ratio=None, max_iter=100,
                                                 multi_class='auto',
                                                 n_jobs=None, penalty='l2',
                                                 random_state=None,
                                                 solver='lbfgs', tol=0.0001,
                                                 verbose=0, warm_start=False),
                    n_jobs=None)

In [50]:
# Make Predictions on Train Set
train_pred_lr=lr_model.predict_proba(train_word_features)

In [51]:
train_pred_lr[:5]

array([[0.02628422, 0.03802301, 0.32864872, 0.0283266 , 0.04047505,
        0.0257426 , 0.7458225 , 0.14566368, 0.06306098, 0.68532804],
       [0.0205731 , 0.02353522, 0.06335822, 0.09140213, 0.03628546,
        0.01972218, 0.49634887, 0.89388967, 0.04181783, 0.56007417],
       [0.0984721 , 0.02163289, 0.03923887, 0.2188406 , 0.30334228,
        0.02699562, 0.52203059, 0.73301885, 0.06461424, 0.03406334],
       [0.01823237, 0.01210113, 0.35977266, 0.02255171, 0.07155586,
        0.00912383, 0.5446768 , 0.37655145, 0.07385593, 0.59889956],
       [0.02696836, 0.00531752, 0.01962782, 0.02888206, 0.07182421,
        0.00483854, 0.90914529, 0.19631622, 0.01220511, 0.97522508]])

In [52]:
# Finding Optimum value
print("Optimal threshold=>",optimum_threshold(y_tr,train_pred_lr))

Optimal threshold=> 0.34


In [53]:
# Getting classes using optimum threshold
train_pred_lr_class=classify(train_pred_lr,0.34)
train_pred_lr_class[:5]

[[0, 0, 0, 0, 0, 0, 1, 0, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 1, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 1, 0, 0],
 [0, 0, 1, 0, 0, 0, 1, 1, 0, 1],
 [0, 0, 0, 0, 0, 0, 1, 0, 0, 1]]

In [54]:
# Evaluating on Training Set
print("F1-score on Train Set:",f1_score(y_tr,train_pred_lr_class,average="weighted"))

F1-score on Train Set: 0.8219911260031991


In [55]:
# Make Predictions on Validation Set
val_pred_lr=lr_model.predict_proba(test_word_features)

# Getting Classes
val_pred_lr_class=classify(val_pred_lr,0.34)

# Evaluating on Validation Set
print("F1-score on Validation Set:",f1_score(y_val,val_pred_lr_class,average="weighted"))

F1-score on Validation Set: 0.7561692360643565


## Model Building Summary
|        Model        | Train Set | Validation Set |
|:-------------------:|:---------:|:--------------:|
|     Naive Bayes     |   0.7273  |     0.6622     |
| Logistic Regression |   0.8220  |     0.7562     |

It is evident from the results that Logistic Regression performs better than Naive Bayes.

# Final Question Tagging Pipeline

In [56]:
def tagging(question):
  # Text Cleaning
  cleaned_question=cleaner(question)
  
  # Feature Engineering
  vector=word_vectorizer.transform([cleaned_question])
  
  # Predicting Probabilities
  pred_prob=lr_model.predict_proba(vector)
  
  # Converting Probabilities into classes
  pred_class=classify(pred_prob,0.34)

  return mlb.inverse_transform(np.array(pred_class))

<font size=4>**Sample Question:**</font>
<p>I'm using SVM classification (Matlab) within my research works, and I want to know:  </p><ol><li>The advantages and disadvantages of each training algorithm, i.e., SMO, LS and QP<br></li><li>In general case, what is the suitable algorithm?<br></li><li>If the choice of one of them can impact the classification performance?<br></li><li>If the are any relationship between the choice of training method (or algorithm) and the choice of the kernel function.</li></ol>

In [57]:
tagging("<p>I'm using SVM classification (Matlab) within my research works, and I want to know:  </p><ol><li>The advantages and disadvantages of each training algorithm, i.e., SMO, LS and QP<br></li><li>In general case, what is the suitable algorithm?<br></li><li>If the choice of one of them can impact the classification performance?<br></li><li>If the are any relationship between the choice of training method (or algorithm) and the choice of the kernel function.</li></ol")

[('classification', 'machine learning')]